src/cmap/auth/mongodb_oidc/callback_lock_cache.ts

```diff
@@ -0,0 +1,135 @@
+import type { MongoCredentials } from '../mongo_credentials';
+import type {
+  IdPServerInfo,
+  IdPServerResponse,
+  OIDCCallbackContext,
+  OIDCRefreshFunction,
+  OIDCRequestFunction
+} from '../mongodb_oidc';
+
+/** Error message for when request callback is missing. */
+const REQUEST_CALLBACK_REQUIRED_ERROR =
+  'Auth mechanism property REQUEST_TOKEN_CALLBACK is required.';
+/* Counter for function "hashes".*/
+let FN_HASH_COUNTER = 0;
+/* No function present function */
+const NO_FUNCTION: OIDCRequestFunction = () => {
+  return Promise.resolve({ accessToken: 'test' });
+};
+/* The map of function hashes */
+const FN_HASHES = new WeakMap<OIDCRequestFunction | OIDCRefreshFunction, number>();
+/* Put the no function hash in the map. */
+FN_HASHES.set(NO_FUNCTION, FN_HASH_COUNTER);
+
+/**
+ * An entry of callbacks in the cache.
+ */
+interface CallbacksEntry {
+  requestCallback: OIDCRequestFunction;
+  refreshCallback?: OIDCRefreshFunction;
+  callbackHash: string;
+}
+
+/**
+ * A cache of request and refresh callbacks per server/user.
+ */
+export class CallbackLockCache {
+  entries: Map<string, CallbacksEntry>;
+
+  /**
+   * Instantiate the new cache.
+   */
+  constructor() {
+    this.entries = new Map<string, CallbacksEntry>();
+  }
+
+  /**
+   * Clear the cache.
+   */
+  clear() {
+    this.entries.clear();
+  }
+
+  /**
+   * Get the callbacks for the connection and credentials. If an entry does not
+   * exist a new one will get set.
+   */
+  getCallbacks(connection: Connection, credentials: MongoCredentials): CallbacksEntry {
+    const requestCallback = credentials.mechanismProperties.REQUEST_TOKEN_CALLBACK;
+    const refreshCallback = credentials.mechanismProperties.REFRESH_TOKEN_CALLBACK;
+    if (!requestCallback) {
+      throw new MongoInvalidArgumentError(REQUEST_CALLBACK_REQUIRED_ERROR);
+    }
+    const callbackHash = hashFunctions(requestCallback, refreshCallback);
+    const key = cacheKey(connection, credentials, callbackHash);
+    const entry = this.entries.get(key);
+    if (entry) {
+      return entry;
+    }
+    return this.setCallbacks(key, callbackHash, requestCallback, refreshCallback);
+  }
+
+  /**
+   * Set locked callbacks on for connection and credentials.
+   */
+  private setCallbacks(
+    key: string,
+    callbackHash: string,
+    requestCallback: OIDCRequestFunction,
+    refreshCallback?: OIDCRefreshFunction
+  ): CallbacksEntry {
+    const entry = {
+      requestCallback: withLock(requestCallback),
+      refreshCallback: refreshCallback ? withLock(refreshCallback) : undefined,
+      callbackHash: callbackHash
+    };
+    this.entries.set(key, entry);
+    return entry;
+  }
+}
+
+/**
+ * Get a cache key based on connection and credentials.
+ */
+function cacheKey(
+  connection: Connection,
+  credentials: MongoCredentials,
+  callbackHash: string
+): string {
+  return `${connection.address}-${credentials.username}=${callbackHash}`;
```

Comment:
Obviously this is not a concern in practice, but what if `me@example.com` authenticates to `cluster.net-bar.net` and `bar.net-me@example.com` authenticates to `cluster.net`?

```suggestion
  return JSON.stringify([connection.address, credentials.username, callbackHash}]);
```

Reply:
I've updated to suggestion in both caches.